src/nupic/algorithms/BacktrackingTMCpp.cpp

```diff
@@ -0,0 +1,1871 @@
+
+          // Loop through all segments in tm2seg and see if any of them match
+          // tm1seg
+          bool res = false;
+          for (Size tm2segIdx = 0; tm2segIdx < nSegs; tm2segIdx++) {
+            BacktrackingTMCpp::SegOnCellInfo_t tm2seg;
+            tm2seg = tm2.getSegmentOnCell(c, i, tm2segIdx);
+            if (sameSegment(tm1seg, tm2seg) == true) {
+              res = true;
+              break;
+            }
+          }
+          if (!res) {
+            result = false;
+            if (verbosity >= 0) {
+              out << "\nSegments are different for cell: [" << c << ", " << i
+                  << "]\n";
+              out << "TM1: ";
+              tm1.printCell(c, i, false, out);
+              out << "TM2: ";
+              tm2.printCell(c, i, false, out);
+            }
+          }
+        }
+      }
+    }
+  }
+  if (result == true) {
+    if (verbosity > 1)
+      out << "TM's match" << std::endl;
+  } else {
+    if (verbosity > 1)
+      out << std::endl;
+  }
+  return result;
+}
+
+// identify the differences between this and another TM.
+// compare parameters and everything that is serializeable.
+// What does it mean for two tm's to be the same?
+// If the two TM's were run in parallel they will have different
+// results because the random number generator gives different numbers.
+// The only way to get the same numbers is the following:
+//    1) Run the first TM to some point.
+//    2) save the TM.
+//    3) Run the first TM to another point.
+//    4) Restore the saved TM.  The restore should restore the Random generator.
+//    5) Run the second TM to the same point using the same data.
+// Now you can use this function to compare the two tm's.
+// The two will be the same.
+
+bool BacktrackingTMCpp::diff(const BacktrackingTMCpp &tm1,
+                             const BacktrackingTMCpp &tm2) {
+  std::stringstream ss1;
+  std::stringstream ss2;
+  ss1.clear();
+  ss2.clear();
+  tm1.save(ss1);
+  tm2.save(ss2);
+  bool same = (ss1.str() == ss2.str());
+  return same;
+}
+
+/////////////////////////////////////////////////////////////
+
+// A segment is active if it has >= activationThreshold connected
+// synapses that are active due to infActiveState. timestep is "t" or "t-1".
+bool BacktrackingTMCpp::_slowIsSegmentActive(Segment &seg,
+                                             const char *timestep) const {
+  Size numActiveSyns = 0;
+  UInt32 threshold = cells4_->getActivationThreshold();
+  for (UInt synIdx = 0; synIdx < (UInt)seg.size(); synIdx++) {
+    if (seg.getPermanence(synIdx) >= cells4_->getPermConnected()) {
+      Size srcIdx = seg.getSrcCellIdx(synIdx);
+      Byte *state = (!strcmp(timestep, "t")) ? cells4_->getInfActiveStateT()
+                                             : cells4_->getInfActiveStateT1();
+      if (state[srcIdx]) {
+        numActiveSyns += 1;
+        if (numActiveSyns >= threshold)
+          return true;
+      }
+    }
+  }
+  
```

Comment:
I wanted to discuss this with you, since it may be relevant to the design of bitstream serialization you are implementing here. 

Back story: I wanted to make (still) the outputType_ const, as it would catch ie. the error we had with NTA_ASSERT(outType = "aaa"). 

Problem: 
> in >> (const)outputType_ 

is of course limiting. 

Message: 

Could the bitstream serialization work as follows? 
<Serializable> API: load/save methods. 
where save is ofc instance method. 
`<T> static void load(...)`  -- load is static, returns the newly constructed object. 
+: no limitations in object's implementation, it's constructed via normal constructor plus extra internal members are set (ie cells4_)
-: we'll have to make some internal variables protected, or make a setup method, so the cells4_ can be assigned the new, loaded value. 

What is your opinion?